### ISSUE TITLE
Add SocketsHttpHandler.PlaintextStreamFilter

### ISSUE THREAD

**Issue by geoffkizer** [dotnet/runtime#42557]
## Background and Motivation

We recently removed System.Net.Connections from 5.0. One feature we lost was the PlaintextFilter property on SocketsHttpHandler, which allowed you to inspect and modify unencrypted (i.e. "plaintext") HTTP protocol traffic in and out.

This new API provides the same capability without relying on System.Net.Connections. When you set this property, your callback will be invoked whenever SocketsHttpHandler constructs a new HTTP connection, after completing the TCP connect and TLS establishment. You can use the callback to wrap the underlying Stream in your own Stream.

## Proposed API

```csharp
namespace System.Net.Http
{
    public sealed class SocketsHttpPlaintextStreamFilterContext
    {
        public Stream PlaintextStream => _plaintextStream;
        public Version HttpVersion => _httpVersion;
        public HttpRequestMessage RequestMessage => _requestMessage;
    }
    public sealed class SocketsHttpHandler : HttpMessageHandler
    {
        public Func<SocketsHttpPlaintextStreamFilterContext, CancellationToken, ValueTask<Stream>>? PlaintextStreamFilter { get; set; }
    }
}
```

See ConnectCallback API for comparison: https://github.com/dotnet/runtime/issues/41949

## Alternative Designs

What's the right name for this feature? In the original Connections design, it was just PlaintextFilter.
Does including "Stream" in the name help?
"Plaintext" vs "Raw"? "Raw" may sound more like the on-the-wire, encrypted bits. "Plaintext" may provide better intuition here but isn't a super common term.
Should the property name include "Callback" for consistency with ConnectCallback?

**Comment by terrajobst** [comment]
* Looks good as proposed, modulo:
* `HttpVersion` we should consider a term that conveys that it's the version after policy, such as `NegotiatedHttpVersion`
* `RequestMessage` should convey that this is the initial message only, such as `InitialRequestMessage`, because the callback is called once per connection, not per request. We should also rename the property on `SocketsHttpConnectionContext`.

```C#
namespace System.Net.Http
{
    public sealed class SocketsHttpPlaintextStreamFilterContext
    {
        public Stream PlaintextStream { get; }
        public Version NegotiatedHttpVersion { get; }
        public HttpRequestMessage InitialRequestMessage { get; }
    }
    public sealed class SocketsHttpHandler : HttpMessageHandler
    {
        public Func<SocketsHttpPlaintextStreamFilterContext, CancellationToken, ValueTask<Stream>>? PlaintextStreamFilter { get; set; }
    }
}
```

**Comment by jeffhandley** [comment]
@geoffkizer Can you create a PR for `master` too please?  For other fixes into RC2, we've been merging into `master` first and then porting into `release/5.0-rc2`, but it's OK to keep #42563 since it's already in place.  That branch's commits won't be automatically ported into `master` though.

**Comment by geoffkizer** [comment]
Will do

**Comment by karelz** [comment]
Fixed in 6.0/master (PR #42660) and in 5.0 (PR #42563)
